src/oauth_helpers.js

```diff
@@ -0,0 +1,222 @@
+  const url = new URL(`${ORBIT_API_ROOT_URL}/oauth/token`);
+  let params = new URLSearchParams({
+    grant_type: "refresh_token",
+    client_id: OAUTH_CLIENT_ID,
+    refresh_token: refreshToken,
+  });
+
+  url.search = params.toString();
+
+  try {
+    const response = await fetch(url, {
+      method: "POST",
+    });
+
+    const { access_token, refresh_token, expires_in } = await response.json();
+
+    // Calculate timestamp when OAuth token expires - current time + it's expires_in timestamp
+    const expiresAt = Math.floor(Date.now() / 1000) + expires_in;
+
+    const items = {
+      accessToken: access_token,
+      refreshToken: refresh_token,
+      expiresAt: expiresAt,
+    };
+
+    chrome.storage.sync.set({ authentication: items });
+
+    return items;
+  } catch (err) {
+    console.error(err);
+
+    // If the request fails (for example if the refresh token has expired),
+    // remove the OAuth credentials from storage. This will let the user
+    // fall back on their API token if present, or ask them to reauthenticate.
+    const items = {
+      accessToken: "",
+      refreshToken: "",
+      expiresAt: -1,
+    };
+
+    chrome.storage.sync.set({ authentication: items });
+
+    return items;
+  }
+}
+
+/**
+ * Parses the query parameters from the search portion of a URL and returns them as an object.
+ * @param {URL} url - The URL to parse.
+ * @returns {Object} An object containing the parsed query parameters.
+ * The keys in the object are the parameter names, and the values are either strings or arrays of strings.
+ * If a parameter appears multiple times, its values will be combined into an array.
+ * If the URL has no search portion, an empty object is returned.
+ */
+export function fetchQueryParams(stringUrl) {
+  try {
+    const url = new URL(stringUrl);
+    const searchParams = new URLSearchParams(url.search);
+    const queryParams = {};
+
+    for (const [key, value] of searchParams) {
+      // If param is new, add it to the object to return
+      if (!queryParams.hasOwnProperty(key)) {
+        queryParams[key] = value;
+        // If existing value is an array, append new value to it
+      } else if (Array.isArray(queryParams[key])) {
+        queryParams[key].push(value);
+        // If existing value is not an array, ocnvert it to an array & add new value
+      } else {
+        queryParams[key] = [queryParams[key], value];
+      }
+    }
+
+    return queryParams;
+  } catch (err) {
+    console.error(err);
+    return {};
+  }
+}
+
+/**
+ * Generates a code_verifier
+ * > a random string called code_verifier is generated using the characters: [A-Z], [a-z], [0-9], "-", ".",
+ * > "_" and "~", with a minimum length of 43 characters and a maximum length of 128 characters
+ * https://doorkeeper.gitbook.io/guides/ruby-on-rails/pkce-flow#how-can-pkce-prevent-this
+ */
+export function generateCodeVerifier() {
+  // Initialise empty array of recommended length
+  const codeVerifierLength = 43;
```

Comment:
Does it work if we set it to 128 characters? If it’s the same to us then I’d suggest we go with hardened security 👷

Reply:
It does indeed :D